SU2_CFD/include/numerics_simd/util.hpp

```diff
@@ -0,0 +1,181 @@
+ * \note These should be used instead of C-style arrays.
+ */
+template<class Type, size_t Size>
+using Vector = C2DContainer<unsigned long, Type, StorageType::ColumnMajor, simd::SIMD_SIZE, Size, 1>;
+
+template<size_t Size> using VectorInt = Vector<Int, Size>;
+template<size_t Size> using VectorDbl = Vector<Double, Size>;
+
+template<class Type, size_t Rows, size_t Cols>
+using Matrix = C2DContainer<unsigned long, Type, StorageType::RowMajor, simd::SIMD_SIZE, Rows, Cols>;
+
+template<size_t Rows, size_t Cols = Rows> using MatrixInt = Matrix<Int, Rows, Cols>;
+template<size_t Rows, size_t Cols = Rows> using MatrixDbl = Matrix<Double, Rows, Cols>;
+
+/*!
+ * \brief Simple pair type for i/j variables.
+ */
+template<class T>
+struct CPair {
+  T i, j;
+};
+
+/*!
+ * \brief Dot product.
+ */
+template<size_t nDim, class ForwardIterator, class T>
+FORCEINLINE Double dot(ForwardIterator iterator, const T* ptr) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += *(iterator++) * ptr[iDim];
+  }
+  return sum;
+}
+
+/*!
+ * \overload Dot product.
+ */
+template<size_t nDim, class ForwardIterator>
+FORCEINLINE Double dot(ForwardIterator iterator, const VectorDbl<nDim>& vector) {
+  return dot<nDim>(iterator, vector.data());
+}
+
+/*!
+ * \overload Dot product.
+ */
+template<size_t nDim>
+FORCEINLINE Double dot(const VectorDbl<nDim>& a, const VectorDbl<nDim>& b) {
+  return dot<nDim>(a.data(), b.data());
+}
+
+/*!
+ * \brief Squared norm.
+ */
+template<size_t nDim, class ForwardIterator>
+FORCEINLINE Double squaredNorm(ForwardIterator iterator) {
+  Double sum = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    sum += pow(*(iterator++),2);
+  }
+  return sum;
+}
+
+/*!
+ * \overload Squared norm.
+ */
+template<size_t nDim>
+FORCEINLINE Double squaredNorm(const VectorDbl<nDim>& vector) {
+  return squaredNorm<nDim>(vector.data());
+}
+
+/*!
+ * \brief Vector norm.
+ */
+template<size_t nDim>
+FORCEINLINE Double norm(const VectorDbl<nDim>& vector) { return sqrt(squaredNorm(vector)); }
+
+/*!
+ * \brief Gather a single variable from index iPoint of a 1D container.
+ */
+template<class Container>
+FORCEINLINE Double gatherVariables(Int iPoint, const Container& vars) {
+  auto x = *vars.innerIter(iPoint);
+  AD::SetPreaccIn(x);
+  return x;
+}
+
+/*!
+ * \brief Gather a vector of variables (size nVar) from row iPoint of a 2D container.
+ */
+template<size_t nVar, class Container>
+FORCEINLINE VectorDbl<nVar> gatherVariables(Int iPoint, const Container& vars) {
+  auto x = vars.template get<VectorDbl<nVar> >(iPoint);
+  AD::SetPreaccIn(x, nVar);
+  return x;
+}
+
+/*!
+ * \brief Gather a matrix of variables from outer index iPoint of a 3D container.
+ */
+template<size_t nRows, size_t nCols, class Container>
+FORCEINLINE MatrixDbl<nRows,nCols> gatherVariables(Int iPoint, const Container& vars) {
+  auto x = vars.template get<MatrixDbl<nRows,nCols> >(iPoint);
+  AD::SetPreaccIn(x, nRows, nCols);
+  return x;
+}
```

Comment:
I'm handling pre-accumulation by registering variables as they are retrieved from the containers, which I think should make the whole process much safer.

Fusing the convective and viscous fluxes reduces the memory used by the AD tape by about 20% for a 2D RANS SST problem. The recording seems to be slightly more expensive but each iteration then runs faster.